File: 2-add-two-numbers/main.go
package main

import (
	"fmt"
	"log"
	"math/big"
	"strconv"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func arrToListNode(arr []int) *ListNode {
	var current, next *ListNode
	for i := len(arr) - 1; i >= 0; i-- {
		v := arr[i]
		current = &ListNode{v, next}
		next = current
	}
	return current
}

func listNodeToBigInt(node *ListNode) *big.Int {
	if node == nil {
		return nil
	}
	val := big.NewInt(int64(node.Val))
	next := listNodeToBigInt(node.Next)
	if next == nil {
		return val
	} else {
		// Multiply the next value by 10, since it is in the next tens place.
		next.Mul(big.NewInt(10), next)

		// Add the result to the current value.
		val.Add(val, next)

		// Pass along the final value.
		return val
	}
}

func Reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

func bigIntToListNode(bi *big.Int) *ListNode {
	s := bi.String()
	reversed := Reverse(s)
	arr := []int{}
	for _, c := range reversed {
		digit, err := strconv.Atoi(string(c))
		if err != nil {
			log.Fatal(err)
		}
		arr = append(arr, digit)
	}
	return arrToListNode(arr)
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	n1 := listNodeToBigInt(l1)
	n2 := listNodeToBigInt(l2)
	val := new(big.Int)
	val.Add(n1, n2)
	return bigIntToListNode(val)
}

func nodesEqual(n1, n2 *ListNode) bool {
	if n1 == nil && n2 == nil {
		return true
	}
	if n1 == nil || n2 == nil {
		return false
	}
	if n1.Val != n2.Val {
		return false
	}
	return nodesEqual(n1.Next, n2.Next)
}

func print_result(name string, res, exp *ListNode) {
	fmt.Printf("%v: ", name)
	if nodesEqual(res, exp) {
		fmt.Println("success")
	} else {
		fmt.Println("failed")
	}
}

func main() {
	var testCases = []struct {
		l1   []int
		l2   []int
		exp  []int
		name string
	}{
		{
			[]int{2, 4, 3},
			[]int{5, 6, 4},
			[]int{7, 0, 8},
			"case 1",
		},
		{
			[]int{0},
			[]int{0},
			[]int{0},
			"case 2",
		},
		{
			[]int{9, 9, 9, 9, 9, 9, 9},
			[]int{9, 9, 9, 9},
			[]int{8, 9, 9, 9, 0, 0, 0, 1},
			"case 3",
		},
	}

	for _, v := range testCases {
		l1 := arrToListNode(v.l1)
		l2 := arrToListNode(v.l2)
		res := addTwoNumbers(l1, l2)
		exp := arrToListNode(v.exp)
		print_result(v.name, res, exp)
	}
}

File: 2-add-two-numbers/main.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def arr_to_list_node(arr: List[int]) -> ListNode:
    current = None
    next = None
    for x in reversed(arr):
        current = ListNode(x, next=next)
        next = current
    return current


def node_to_int(node: ListNode) -> int:
    if node.next is None:
        return node.val

    next = node_to_int(node.next) * 10
    value = node.val
    return value + next


def int_to_node(num: int) -> ListNode:
    s = reversed(str(num))
    return arr_to_list_node([int(x) for x in s])


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # This works because Python 3 uses long ints, which are limited only by
        # memory: https://www.python.org/dev/peps/pep-0237/
        l1_int = node_to_int(l1)
        l2_int = node_to_int(l2)
        sum = l1_int + l2_int
        return int_to_node(sum)


if __name__ == '__main__':
    s = Solution()

    for l1, l2, exp, name in [
        (
            [2, 4, 3],
            [5, 6, 4],
            [7, 0, 8],
            "case 1",
        ),
        (
            [0],
            [0],
            [0],
            "case 2",
        ),
        (
            [9, 9, 9, 9, 9, 9, 9],
            [9, 9, 9, 9],
            [8, 9, 9, 9, 0, 0, 0, 1],
            "case 3",
        ),
    ]:
        res = s.addTwoNumbers(arr_to_list_node(l1), arr_to_list_node(l2))
        if node_to_int(res) != node_to_int(arr_to_list_node(exp)):
            print(f"{name}, exp: {exp}, res: {res}")
        else:
            print(f"passed: {name}")
